src/Gui/Application.cpp

```diff
@@ -636,19 +637,21 @@ void Application::open(const char* FileName, const char* Module)
                 }
             }
             else {
-                // issue module loading
-                Command::doCommand(Command::App, "import %s", Module);
-
-                // check for additional import options
-                std::stringstream str;
-                str << "if hasattr(" << Module << ", \"importOptions\"):\n"
-                    << "    options = " << Module << ".importOptions(u\"" << unicodepath << "\")\n"
-                    << "    " << Module << ".open(u\"" << unicodepath << "\", options = options)\n"
-                    << "else:\n"
-                    << "    " << Module << ".open(u\"" << unicodepath << "\")\n";
-
-                std::string code = str.str();
-                Gui::Command::runCommand(Gui::Command::App, code.c_str());
+                // Load using provided python module
+                {
+                    Base::PyGILStateLocker locker;
+                    Py::Module moduleIo(PyImport_ImportModule("freecad.module_io"));
+                    const auto dictS = moduleIo.getDict().keys().as_string();
+                    if (!moduleIo.isNull() && moduleIo.hasAttr("OpenInsertObject"))
+                    {
+                        const Py::TupleN args(
+                            Py::Module(PyImport_ImportModule(Module)),
+                            Py::String(unicodepath),
+                            Py::String("open")
+                        );
+                        moduleIo.callMemberFunction("OpenInsertObject", args);
+                    }
+                }
```

Comment:
Pulling out the doCommand and runCommand here means that the actions taken will no longer be logged to the python console, and can no longer be recorded in a macro.  Is that intentional?

Reply:
You're right, and that wasn't intentional. I wrote it that way because @wwmayer above said to use the PyCXX API - I didn't see a way to square that with using the `Gui::Command` class. Is there a way to have structured C++ code (e.g. using PyCXX rather than mashing strings together) and also have the commands visible in the console/macroable? Should we revert this part of the change? I had a cleaned up version of the string construction that used `QString::arg` in a prior commit - that wouldn't have this regression.

https://github.com/FreeCAD/FreeCAD/blob/de122b6dfa1555b214bb9eaece6153d832109d02/src/Gui/Application.cpp#L601-L627

(yes I messed up the exception name here)

Or do we not care? Or is there another way?

Reply:
Well, I added doConsoleSkip recently to support cases where you want to put something in the console/macro that is a python equivalent without executing it.  But you need to have the python equivalent available, so in this case you would likely need to move back towards the python.  I can't answer the do we care question - I think so, but maybe we're moving in a different direction?